### PR TITLE
providers/vsphere: added host_system for vm deployment

### PR DESCRIPTION
**Pull request by tommynsong**
i have a need to be able to define an esxi host to clone a VM to and hoping that this PR could be of use for others with this niche need.

### CODE REVIEW

**Comment by chrislovecnm** [comment]
@tommynsong thanks for the work.  Nice to meet you, and we have a developers vsphere slack channel.  Ping me if you are interested.

A couple of notes:
- please squash your commits
- please provide unit tests
- please provide output of all unit tests running locally

If you need help please ask!

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go [hunk 1]
```diff
@@ -163,6 +164,12 @@ func resourceVSphereVirtualMachine() *schema.Resource {
 				ForceNew: true,
 			},
 
+			"host_system": &schema.Schema{
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: Does this need validation of any kind?

File: website/source/docs/providers/vsphere/r/virtual_machine.html.markdown [hunk 2]
```diff
@@ -68,6 +68,7 @@ The following arguments are supported:
 * `datacenter` - (Optional) The name of a Datacenter in which to launch the virtual machine
 * `cluster` - (Optional) Name of a Cluster in which to launch the virtual machine
 * `resource_pool` (Optional) The name of a Resource Pool in which to launch the virtual machine. Requires full path (see cluster example).
+* `host_system` (Optional) The name of a ESXi host in which to launch the virutal machine. Requires FQDN as presented in vCenter.
```
Comment (chrislovecnm) [inline website/source/docs/providers/vsphere/r/virtual_machine.html.markdown]: Would `esxi_host` make more sense?

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go [hunk 3]
```diff
@@ -1488,9 +1505,34 @@ func (vm *virtualMachine) setupVirtualMachine(c *govmomi.Client) error {
 	}
 
 	var resourcePool *object.ResourcePool
-	if vm.resourcePool == "" {
-		if vm.cluster == "" {
-			resourcePool, err = finder.DefaultResourcePool(context.TODO())
+	var hostSystem *object.HostSystem
+	if vm.hostSystem == "" {
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: This appears to be repeated code, would it make sense to encapsulate this logic into an interface?

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go [hunk 4]
```diff
@@ -1488,9 +1505,34 @@ func (vm *virtualMachine) setupVirtualMachine(c *govmomi.Client) error {
...
+				resourcePool, err = finder.DefaultResourcePool(context.TODO())
+				if err != nil {
+					return err
+				}
+				hostSystem, err = finder.DefaultHostSystem(context.TODO())
+				if err != nil {
+					return err
+				}
+			} else {
+				resourcePool, err = finder.ResourcePool(context.TODO(), "*"+vm.cluster+"/Resources")
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: Are we using `vm.cluster+"/Resources` on other `finders`, or is this specific to this use case?

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go [hunk 5]
```diff
@@ -1488,9 +1505,34 @@ func (vm *virtualMachine) setupVirtualMachine(c *govmomi.Client) error {
...
+				}
+				hostSystem, err = finder.HostSystemOrDefault(context.TODO(), "*"+vm.cluster)
+				if err != nil {
+					return err
+				}
+			}
+		} else if vm.cluster == "" {
+			resourcePool, err = finder.ResourcePool(context.TODO(), vm.resourcePool)
+			if err != nil {
+				return err
```
Comment (chrislovecnm) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: Please add some logging on `err`s, if it makes set reformat the `err` as well.

**Comment by tommynsong** [comment]
Hi Chris,

I would love to join the channel.
Please send invite to tommynsong@gmail.com
I will review your comments on GH and adjust accordingly, thank you for
your advice.

File: website/source/docs/providers/vsphere/r/virtual_machine.html.markdown (hunk 2 above)
Comment (tommynsong) [inline website/source/docs/providers/vsphere/r/virtual_machine.html.markdown]: in govmomi's object, it is being referred to as HostSystem, hence i decided to follow suite.

File: builtin/providers/vsphere/resource_vsphere_virtual_machine.go (hunk 4 above)
Comment (tommynsong) [inline builtin/providers/vsphere/resource_vsphere_virtual_machine.go]: it is in the original code to construct the string for path to the default resource pool a the cluster.

**Comment by stack72** [comment]
Ping @tommynsong and @chrislovecnm for updates on this? :)

**Comment by chrislovecnm** [comment]
@tommynsong do you need to rebase this?  I think we have some open comments as well.

**Comment by mrjcleaver** [comment]
Did @tommynsong get added to the Slack channel?

**Comment by fredead** [comment]
Im interested in this patch getting merged. If it needs some help getting along please ping me.

**Comment by tommynsong** [comment]
@fredead if you can help getting this along that would be great help. i am currently buried with multiple projects and hardly have any time to work on this patch.

**Comment by fredead** [comment]
I have some tidy ups mainly around finder searches making the code a little neater. 
Im looking into tests although for various reasons these are difficult to run. 

Q: How to get the changes here ?  opening a new PR Is easier for me but seems wrong

**Comment by tommynsong** [comment]
@fredead if a new PR is easier for you, feel free to. i would rather having this feature got merged than my own glory :)

**Comment by Freyert** [comment]
@tommynsong this provider has been moved to https://github.com/terraform-providers/terraform-provider-vsphere

**Comment by apparentlymart** [comment]
Hello @tommynsong, and thanks for working on this!

As part of the the Terraform 0.10 release earlier this year, all of the Terraform providers were moved to their own repositories in [the terraform-providers GitHub organization](https://github.com/terraform-providers), and removed from the Terraform Core repository.

Unfortunately due to the fact that new issues and pull requests are being opened constantly, it was not possible for the various provider maintainers to merge all outstanding pull requests before this split, and there is no automatic way to migrate a pull request to a new repository.

As a result, this pull request can sadly no longer be applied as-is, and so I'm going to close it.

[The vSphere provider](https://github.com/terraform-providers/terraform-provider-vsphere) -- and, in particular, the virtual machine resource -- are currently in the process of being revamped somewhat, with the goal of supporting more of the available features. Unfortunately this means that to re-target this change to the new repository may require some restructuring :confounded:.

Thanks again for working on this, and sorry it was not able to be merged before the provider repository changes.

**Pull request closed without merging**